---
title: "Visualisation"
output: html_notebook
---
CLN3 rank in current annotation
```{r}
trans_count <- read.table("C:/Users/Haoyu/Desktop/Ensembl_110_gene_transcript_count.txt", header = T, sep = "\t")


test <- trans_count %>% filter(Transcript.count >= 50)


#rank CLN3
cln3_rank <- rank(-trans_count$Transcript.count, ties.method = "min")[trans_count$Gene.name == "CLN3"]


percentile_plot <- ggplot(trans_count, aes(x = Transcript.count)) +
  geom_histogram(binwidth = 1, fill = "blue", color = "black") +
  geom_vline(xintercept = trans_count$Transcript.count[trans_count$Gene.name == "CLN3"], color = "red", linetype = "dashed") +
  labs(x = "Transcript Count", y = "Gene Count") + 
  theme(panel.border = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line.x = element_line(size = 0.5, linetype = "solid", colour = "black"),
        axis.line.y = element_line(size = 0.5, linetype = "solid", colour = "black")) +
  ylim(0,4000)


print(percentile_plot)

```



plot exon-exon junctions
```{r fig.height=7, fig.width=16.5}
library(ggplot2)
library(ggtranscript)


annot <- rtracklayer::import("E:/reference_for_APTARS/gencode.v26.annotation.gtf")

annot <- as.data.frame(annot)

# annot <- annot %>%
# separate(transcript_id, sep = "\\.", into = c("transcript_stable_id", "transcript_version"), remove = F)

readthrough <- annot %>% filter(gene_id == "ENSG00000261832.6")

readthrough <- readthrough %>%
  separate(transcript_id, sep = "\\.", into = c("transcript_stable_id", "transcript_version"), remove = F)


readthrough <- readthrough %>%
  separate(gene_id, sep = "\\.", into = c("gene_stable_id", "gene_version"), remove = F)
readthrough <- readthrough %>% filter(transcript_stable_id %in% c("ENST00000635887",
                                               "ENST00000637378", 
                                               "ENST00000637376", 
                                               "ENST00000636078",
                                               "ENST00000637745",
                                               "ENST00000568224",
                                               "ENST00000636503",
                                               "ENST00000636017", 
                                               "ENST00000636866", 
                                               "ENST00000637299", 
                                               "ENST00000638036", 
                                               "ENST00000636766"))

readthrough <- readthrough %>% mutate(junction_group = ifelse(transcript_stable_id == "ENST00000635887", 1 , 
                                                              ifelse(transcript_stable_id == "ENST00000637378", 2,
                                                                     ifelse(transcript_stable_id == "ENST00000636766", 5,
                                                                            ifelse(transcript_stable_id %in% c("ENST00000637376", "ENST00000636078", "ENST00000636503", "ENST00000637745", "ENST00000568224"), 3, 4)))))




junction <- readthrough %>% dplyr::select(transcript_stable_id, junction_group)
junction <- junction %>% mutate(start = ifelse(junction_group %in% c( 1, 4), 28476250, 
                                               ifelse(junction_group == 2, 28482104, 
                                                      ifelse(junction_group == 3, 28477463, 
                                                             ifelse(junction_group == 5, 28477011, NA)))))


junction <- junction %>% mutate(end = ifelse(junction_group %in% c( 1, 2, 3, 5), 28466903, 28471175))
                                               

junction$seqnames <- "chr16"
junction$strand <- "-"

junction <- junction %>% mutate(mean_reads = ifelse(junction_group == 1, 1.298, 
                                               ifelse(junction_group == 2, 1.298, 
                                                      ifelse(junction_group == 3, 2.121, 
                                                             ifelse(junction_group == 4, 1.231, 1.123)))))

junction <- unique(junction)



readthrough_ex <- readthrough %>% filter(type == "exon")
readthrough_cds <- readthrough %>% filter(type == "CDS")

# 
# test <- junction[2,]
# 
# test <- test[rep(seq_len(nrow(test)), each = 17),]
# test_ex <- readthrough_ex %>% filter(transcript_stable_id %in% test$transcript_stable_id)
# test_cds <- readthrough_cds %>% filter(transcript_stable_id %in% test$transcript_stable_id)
# test_ex %>% 
#   ggplot(aes(
#   xstart = start,
#   xend = end,
#   y = reorder(transcript_stable_id, -junction_group),
#   
# )) +
#   geom_range(
#     # fill = "white",
#     height = 0.25
#   ) +
#   geom_range(
#     data = test_cds,
#     aes(fill = gene_stable_id)
#   ) +
#   geom_intron(
#     data = to_intron(test_ex, "transcript_stable_id"),
#     aes(strand = strand),
#     arrow.min.intron.length = 500,
#   ) + 
#   geom_junction(
#     data = test,
#     aes(size = mean_reads),
#     junction.y.max = 0.0013
#   ) +
#   geom_junction_label_repel(
#     data = test,
#     aes(label = round(mean_reads, 2)),
#     #aes(label = junction_group),
#     junction.y.max = 0.0013
#   ) +
#   scale_size_continuous(range = c(0.1,1))+
#   theme_bw()+ 
#   xlab("Coordinates on CHR16") + 
#   ylab(NULL) + 
#   scale_fill_manual(values = c("ENSG00000261832" = "#00BFC4"))+
#   #guides(fill = guide_legend(ncol = 1))+
#   theme(axis.title.x = element_text(face = "bold"),
#         axis.title.y = element_text(face = "bold"))
# 


readthrough_junction <- readthrough_ex %>% 
  ggplot(aes(
  xstart = start,
  xend = end,
  y = reorder(transcript_stable_id, -junction_group),
  
)) +
  geom_range(
    # fill = "white",
    height = 0.25
  ) +
  geom_range(
    data = readthrough_cds,
    aes(fill = gene_stable_id)
  ) +
  geom_intron(
    data = to_intron(readthrough_ex, "transcript_stable_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + 
  geom_junction(
    data = junction,
    aes(size = mean_reads),
    junction.y.max = 0.0015
  ) +
  # geom_junction_label_repel(
  #   data = test,
  #   aes(label = round(mean_reads, 2)),
  #   #aes(label = junction_group),
  #   junction.y.max = 0.0013
  # ) +
  scale_size_continuous(range = c(0.1,1))+
  theme_bw()+ 
  xlab("Coordinates on CHR16") + 
  ylab(NULL) + 
  scale_fill_manual(values = c("ENSG00000261832" = "#00BFC4"))+
  #guides(fill = guide_legend(ncol = 1))+
  theme(axis.title.x = element_text(face = "bold"),
        axis.title.y = element_text(face = "bold"))


readthrough_junction
```



plot genes
```{r fig.height=1.5, fig.width=10}
library(Gviz)
genes <- import.bed("E:/Encode_annotation/gene_annot.bed")

genes <- as_tibble(genes)


genes_track <- AnnotationTrack(genes, name = "Genes", chromosome = "chr16", feature = c("CLN3", "NPIPB7", "CLN3_readthrough"))

genomeAxis <- GenomeAxisTrack(name="MyAxis", add35=T) 



plotTracks(list(genes_track, genomeAxis), 
           CLN3= "#F8766D", 
           CLN3_readthrough="#00BFC4",
           NPIPB7 = "darkgreen",
           chromosome = "chr16",
           from=28455000,to=28500000, 
            title.width = 1, 
           cex.title = 0.9, 
           sizes = c(7,3), 
           background.panel = "white", 
           background.title = "darkblue")
```

Plot representative transcripts
```{r fig.height=1.5, fig.width=15}
fusion_whole_gtf <- import("E:/Encode_annotation/all/locus_fusion_whole_gtf.gtf") %>%
  as_tibble()

fusion_whole_gtf$gene_name <- "CLN3_readthrough"

select_gtf <- fusion_whole_gtf %>% filter(transcript_id == "ENST00000637376.1") %>% unique()




select_cln3_gtf <- my_id_whole_gtf %>% filter(transcript_id == "CLN3_24_438aa_5UTR_136_3UTR_49")


gtf_to_plot <- bind_rows(select_cln3_gtf,select_gtf) %>% unique() %>% filter(type == "exon")
cds_to_plot <- bind_rows(select_cln3_gtf,select_gtf) %>% unique() %>% filter(type == "CDS")

 gtf_to_plot %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = transcript_id,
  
)) +
  geom_range(
    # fill = "white",
    height = 0.25
  ) +
  geom_range(
    data = cds_to_plot,
    aes(fill = gene_name)
  ) +
  geom_intron(
    data = to_intron(gtf_to_plot, "transcript_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + 
   theme_bw()+ 
   xlab("Coordinates on CHR16") + 
   ylab(NULL)+
   scale_fill_manual(values = c(CLN3= "#F8766D", 
           CLN3_readthrough="#00BFC4")) +
   theme(legend.position = "none",
         axis.title = element_text(face = "bold"))
  

```

check proportion of reads that was mapped to the readthrough gene
```{r}


count <- quanti_info %>% dplyr::select(annot_gene_name,read_counts, tissue)

count_full_tissue <- list()
for (i in 1:length(organs)) {
  tissues <- quanti_tissue %>% filter(organ == organs[i])
  df <- count %>% dplyr::filter(tissue %in% tissues$Biosample.term.name)
  df$organ <- organs[i]
  count_full_tissue[[i]] <- df
}

count <- bind_rows(count_full_tissue)


count <- count %>% dplyr::select(-tissue)

#count <- count %>% group_by(annot_gene_name, organ) %>% summarise_each(list(sum))
count <- count %>% group_by(annot_gene_name, organ) %>% summarise_each(list(mean))
count$organ <- gsub("_", " ", count$organ)

count$annot_gene_name <- gsub("AC138894.1", "CLN3-NPIPB7 readthrough", count$annot_gene_name)

count <- count %>% 
  group_by(organ) %>%
  mutate(plot_rank = cur_group_id())

mis_annot <- count %>% 
  ggplot(aes(x=reorder(organ, -plot_rank), y=read_counts, fill= annot_gene_name)) + 
  geom_bar(position ="fill" , stat = "identity")+ 
  xlab("Organs") + 
  ylab("Read counts proportion") + 
  #ggtitle("Proportion of CLN3 reads mapped to the readthrough gene") + 
  coord_flip() + 
  theme_bw() + 
  guides(fill=guide_legend(title="Gene")) + 
  theme(axis.title = element_text(face = "bold")) + 
  scale_y_continuous(labels = scales::percent)

mis_annot


```



```{r fig.height=6.5, fig.width=10}
library(ggbreak)

sample_depth_comb <- bind_rows(sample_depth_cln3,sample_depth_fusion)


sample_depth_comb$organ <- gsub("_", " ", sample_depth_comb$organ)

exp_both_genes <- sample_depth_comb %>% ggplot(aes(
 x = organ,
 y = TPM,
fill = gene
 )) + geom_boxplot(outlier.shape = NA, width = 0.7)+
xlab("Organs") + 
  ylab("TPM") + 
  theme_bw() +
  scale_color_brewer(palette="Set1") + 
  facet_wrap( ~ gene, scales = "free") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.6, vjust = 0.6),
        strip.text = element_text(face = "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title = element_text(face = "bold", size = 13.5))+
  guides(fill = guide_legend(title = "Gene"))


exp_both_genes
```



Visualise number of transcripts within different transcript types
```{r}
library(RColorBrewer)
n_types <- length(unique(trans_merged_filtered$transcript_type))
qual_col_pals <- brewer.pal.info[brewer.pal.info$category == "qual", ]
col_vector <- unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

types <- unique(trans_merged_filtered$transcript_type)
color_types <- col_vector[60:(60 + n_types - 1)]
color_types[5] <- "#99CC33"

for (i in 1:length(color_types)) {
  names(color_types)[i] <- types[i]
}



novelty_bar <-trans_merged_filtered %>% 
  ggplot(aes(transcript_type, ..count.., fill = transcript_type)) + 
  geom_bar() + 
  xlab("Transcripts types")+ 
  ylab("Counts") +
  theme_classic() +
  scale_fill_manual(values = color_types) + 
  theme(axis.text.x = element_text(angle = 50, vjust = 0.85, hjust = 0.8),
      axis.title.y = element_text(face = "bold", size = 15),
      axis.title.x = element_text(face = "bold", size = 15),
      legend.position = "right") +  
  guides(fill = guide_legend(title = "Transcripts types")) 
  
novelty_bar <- novelty_bar + scale_y_continuous(breaks =c(5,10,15,20,25,30,35,40,45,50,55,60))  




novelty_bar + geom_bar(data = trans_merged_filtered[(trans_merged_filtered$transcript_type %in% c("Coding_Novel_ORF_and_UTR", "Coding_Novel_ORF_only")),],
                       aes(fill = transcript_type),
                       alpha = 0, 
                       size = 1, 
                       color = "black") + guides(fill = guide_legend(override.aes = list(color = NULL)))
```



Visualise all valid transcripts
```{r fig.height=200, fig.width=25}
library(ggtranscript)

library(cowplot)

###just read the file
my_id_whole_gtf <- import("E:/Encode_annotation/all/locus_selection_CLN3_my_id_with_cds.gtf", format = "gtf") %>% as_tibble()
my_id_gtf <- my_id_whole_gtf %>% filter(type == "exon")
my_id_cds <- my_id_whole_gtf %>% filter(type == "CDS")


my_id_gtf$occurrence <- trans_merged_quanti$occurrence[match(my_id_gtf$transcript_id, trans_merged_quanti$new_id)]
my_id_cds$occurrence <- trans_merged_quanti$occurrence[match(my_id_cds$transcript_id, trans_merged_quanti$new_id)]



my_id_cds_filtered <- my_id_cds %>% filter(transcript_id %in% trans_merged_filtered$new_id)
my_id_gtf_filtered <- my_id_gtf %>% filter(transcript_id %in% trans_merged_filtered$new_id)

#plot transcripts with occurrence > 3

structure_filtered <- my_id_gtf_filtered %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = reorder(transcript_id, occurrence),
  
)) +
  geom_range(
    # fill = "white",
    height = 0.25
  ) +
  geom_range(
    data = my_id_cds_filtered,
    aes(fill = transcript_status)
  ) +
  geom_intron(
    data = to_intron(my_id_gtf_filtered, "transcript_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + theme_bw()+ xlab("Coordinates on CHR16") + ylab(NULL)


structure_filtered


structure_full <- my_id_gtf %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = reorder(transcript_id, occurrence),
  
)) +
  geom_range(
    # fill = "white",
    height = 0.25
  ) +
  geom_range(
    data = my_id_cds,
    aes(fill = transcript_status)
  ) +
  geom_intron(
    data = to_intron(my_id_gtf, "transcript_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + theme_bw()+ xlab("Coordinates on CHR16") + ylab(NULL)

structure_full
```

Set colors for ORFs

```{r}
library(RColorBrewer)
###generate n most distinctive colors
trans_merged_filtered$orf_id <- info$orf_id[match(trans_merged_filtered$new_id, info$new_id)]

n <- length(unique(trans_merged_filtered$orf_id))
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))



orf_more_than_3 <- unique(trans_merged_filtered$orf_id)
color_used <- col_vector[12:(n+11)]

pie(rep(1,n), col=sample(color_used, n))


for (i in 1:length(color_used)){
  
  names(color_used)[i] <- orf_more_than_3[i]
  
}


#color_used[26] <- "#669966"
color_used[26] <- "#CC99FF"

```

Plot ORFs usage
```{r fig.height=20, fig.width=10}

orf_usage_full <- ggplot(orf_prop_long, aes(reorder(x=orf_id, occurrence), y= value, fill = orf_id)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(color="black", size=0.9, alpha=0.9, width = 0.25) +
  scale_y_continuous(n.breaks = 9, labels = scales::percent, limits = c(0, 1)) +
  ggtitle("All Tissues") +
  ylab("Proportion") +
  xlab("ORFs") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text = element_text(size = 11),
        title = element_text(face = "bold"),
        plot.title = element_text(size=24), 
        legend.position = "bottom") +
  coord_flip() +
  scale_fill_manual(values = color_used) +
  guides(fill = guide_legend(title = "ORFs", nrow = 5))



orf_usage_full
```


Differential ORF expression
```{r fig.height=4.5, fig.width=7}
quanti_organ <- info %>% dplyr::select(orf_id, tissue, grep('rep', colnames(info)))

organ_info <- read.csv("E:/Encode_annotation/all/organ_info.csv", sep = ",")

#for same tissue types within different organs
orf_organ <- list()

for (i in 1:length(organs)) {
  tissues <- quanti_tissue %>% filter(organ == organs[i])
  df <- quanti_organ %>% dplyr::filter(tissue %in% tissues$Biosample.term.name)
  df$organ <- organs[i]
  orf_organ[[i]] <- df
}

quanti_organ <- bind_rows(orf_organ)



quanti_organ <- quanti_organ %>% dplyr::select(-tissue) %>% group_by(orf_id, organ) %>% summarise_each(list(sum))


#select top 15 ORFs by occurrence

top_15_orf <- orf_prop_sum[1:15,]$orf_id

quanti_organ_top <- quanti_organ %>% 
  ungroup() %>% 
  filter(quanti_organ$orf_id %in% top_15_orf)



quanti_organ_top <- reshape2::melt(quanti_organ_top) %>%
  filter(value != 0)

quanti_organ_top <- quanti_organ_top %>% 
  dplyr::select(-variable) %>% 
  group_by(orf_id, organ) %>% 
  summarise_each(list(sum))  

#reorder
quanti_organ_top <- quanti_organ_top %>%
  group_by(orf_id) %>%
  mutate(mean_usage = mean(value)) %>%
  arrange(desc(-mean_usage))

#quanti_organ_top$orf_id2 <- relevel(as.factor(quanti_organ_top$orf_id), "CLN3_24_438aa")

#rank organ by CLN3_24_438aa usage
organ_rank <- quanti_organ_top %>% 
  group_by(organ) %>%
  mutate(scaled_usage = value / sum(value)) %>%
  filter(orf_id == "CLN3_24_438aa") %>%
  arrange(desc(scaled_usage)) %>%
  ungroup() %>%
  mutate(organ_rank = row_number())

quanti_organ_top$organ_rank <- organ_rank$organ_rank[match(quanti_organ_top$organ, organ_rank$organ)]

quanti_organ_top$orf_id <- factor(quanti_organ_top$orf_id, levels = unique(quanti_organ_top$orf_id[order(quanti_organ_top$value)]))


quanti_organ_top$organ <- gsub("_", " ", quanti_organ_top$organ)

#quanti_organ_top$orf_id <- factor(quanti_organ_top$orf_id, levels = rev(levels(quanti_organ_top$orf_id)))

dou_plot <- ggplot(data = quanti_organ_top, aes(x=reorder(organ, -organ_rank), y=value, fill= orf_id)) + 
  geom_bar(position = "fill", stat = "identity")+
  xlab("Organs") + 
  ylab("ORFs usage") + 
  coord_flip()+
  theme_bw()+
  scale_fill_manual(values = color_used) + 
  scale_y_continuous(labels = scales::percent) + 
  guides(fill=guide_legend(title="ORF IDs")) + 
  theme(axis.title = element_text(face = "bold"))


dou_plot
```

Top15 ORFs usage across all tissues
```{r fig.height=5.5, fig.width=6}
orf_prop_long_top <- orf_prop_long %>% filter(orf_id %in% top_15_orf)

orf_usage_full_top <- ggplot(orf_prop_long_top, aes(reorder(x=orf_id, occurrence), y= value, fill = orf_id)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(color="black", size=0.5, alpha=0.9, width = 0.25) +
  scale_y_continuous(n.breaks = 9, 
                     labels = scales::percent, 
                     limits = c(0,1) )+
  ggtitle("All Tissues") +
  ylab("Proportion") + 
  xlab("ORFs")  +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text = element_text(size = 11),
        title = element_text(face = "bold"),
        plot.title = element_text(size=24), 
        legend.position = "bottom")+
  coord_flip()+ 
  scale_fill_manual(values = color_used)+
  guides(fill = guide_legend( title = "ORF IDs", nrow = 4))

orf_usage_full_top

```





select transcripts to show
```{r fig.height=6, fig.width=22}
my_id_gtf <- rtracklayer::import("E:/Encode_annotation/all/locus_selection_CLN3_my_id_with_cds.gtf") %>% as.data.frame() %>% unique()
my_id_cds <- my_id_gtf %>% filter(type == "CDS") %>% as.data.frame() %>% unique()

#only keep exons for gtf
my_id_gtf <- my_id_gtf %>% filter(type == "exon")


my_id_gtf$occurrence <- trans_merged_quanti$occurrence[match(my_id_gtf$transcript_id, trans_merged_quanti$new_id)]
my_id_cds$occurrence <- trans_merged_quanti$occurrence[match(my_id_cds$transcript_id, trans_merged_quanti$new_id)]


my_id_gtf$orf_id <- info$orf_id[match(my_id_gtf$transcript_id, info$new_id)]
my_id_cds$orf_id <- info$orf_id[match(my_id_cds$transcript_id, info$new_id)]


trans_rank <- transcript_info #%>% filter(occurrence >= 3) 
trans_rank <- trans_rank[with(trans_rank, order(-trans_rank$occurrence)),]
trans_rank <- trans_rank[1:15,]




my_id_cds_top15 <- my_id_cds %>% filter(transcript_id %in% trans_rank$new_id)
my_id_gtf_top15 <- my_id_gtf %>% filter(transcript_id %in% trans_rank$new_id)




get_locus <- function(seqnames, start, end, strand, gene_id) {
  
  locus_subset <- 
    GRanges(seqnames = seqnames,
            ranges = IRanges(start = start, 
                             end = end),
            strand = strand)
}



#zoomed in for top 15
cln3_locus <- get_locus("chr16", 28474111,28495575,"-", "ENSG00000188603" )

structure_filtered_top <- my_id_gtf_top15 %>% 
  ggplot(aes(
  xstart = start,
  xend = end,
  y = reorder(transcript_id, occurrence),
  
)) +
  geom_range(
    # fill = "white",
    height = 0.25
  ) +
  geom_range(
    data = my_id_cds_top15,
    aes(fill = orf_id)
  ) +
  geom_intron(
    data = to_intron(my_id_gtf_top15, "transcript_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + 
  theme_bw()+ 
  xlab("Coordinates on CHR16") + 
  ylab(NULL) + 
  scale_fill_manual(values = color_used)+
  guides(fill = guide_legend(ncol = 1))+
  theme(axis.title.x = element_text(face = "bold"),
        axis.title.y = element_text(face = "bold"))+ 
  xlim(start(cln3_locus), 
       end(cln3_locus))



filtered_prop <- trans_merged_filtered %>% dplyr::select(new_id, colnames(trans_merged_filtered)[grep('rep', colnames(trans_merged_filtered))])
orf_filtered_long <- reshape2::melt(filtered_prop)
orf_filtered_long <- orf_filtered_long %>% dplyr::filter(variable != 'occurrence')
orf_filtered_long$occurrence <- trans_merged_filtered$occurrence[match(orf_filtered_long$new_id, trans_merged_filtered$new_id)]



orf_filtered_long[orf_filtered_long == 0] <- NA
orf_filtered_long <- orf_filtered_long[complete.cases(orf_filtered_long),]

orf_long_top <- orf_filtered_long %>% filter(orf_filtered_long$new_id %in% trans_rank$new_id)


orf_long_top$orf_id <- info$orf_id[match(orf_long_top$new_id, info$new_id)]

usage_top <- ggplot(orf_long_top, aes(reorder(x=new_id, occurrence), y= value, fill = orf_id)) +
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(color="black", size=0.5, alpha=0.9, width = 0.25) +
  scale_y_continuous(n.breaks = 9, labels = scales::percent, limits = c(0,1) )+
  ylab("Proportion") + 
  xlab("Transcripts") +
  theme(plot.title = element_text(size=11),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 1),
        axis.title.x = element_text(face = "bold"),
        axis.title.y = element_text(face = "bold")) +
  coord_flip() + 
  scale_fill_manual(values = color_used) + 
  guides(fill = guide_legend(ncol = 1))



orf_filtered_top <- trans_merged_filtered %>% 
  filter(new_id %in% trans_rank$new_id)

occurrence_top <-  ggplot(orf_filtered_top,
                         aes(x=reorder(new_id, occurrence), 
                             y=occurrence, fill = transcript_type)) +
  geom_col( position = "identity") + 
  theme(axis.text.x = element_text(angle = 90))+ coord_flip() + 
  xlab("Transcript ID") + 
  ylab('Occurrence') + 
  scale_y_continuous(breaks = c(0,20,40,60,80,sample_size)) + 
  theme_bw() + 
  scale_fill_manual(values = color_types) + 
   theme(axis.title.y = element_text(size = rel(1.5), face = "bold"),
         axis.title.x = element_text(size = rel(1), face = "bold"),
        axis.text.x = element_text(size = 11))

 
 
fig2_top <- cowplot::plot_grid(occurrence_top + 
                                 theme(legend.position = "none"), 
                               structure_filtered_top + 
                                 theme(axis.text.y = element_blank(),
                                       axis.ticks.y = element_blank(),
                                       axis.title.y = element_blank(), 
                                       legend.position = "none"), 
                               usage_top + 
                                 theme(axis.title.y = element_blank(),                                                    
                                       axis.text.y = element_blank(),
                                       legend.justification = c(0,0.5),
                                       legend.position = "right"),
                               nrow=1,
                               labels=NA, 
                               rel_widths = c(1.2, 2.2, 1.2), 
                               align = "h")



cage_panel <- plot_grid(NULL, CAGE_plot,NULL, rel_widths = c(1.2, 2.2, 1.2),labels=NA, nrow = 1)
pas_panel <- plot_grid(NULL, pas_plot,NULL, rel_widths = c(1.2, 2.2, 1.2),labels=NA, nrow = 1)

#CAGE_plot and pas_plot created in validation script

fig2 <- plot_grid(fig2_top, cage_panel, pas_panel, nrow = 3, rel_heights = c(5,0.8,0.8))

#cage_panel, and pas_panel from validation.R
fig2
```

Compare major disease transcripts to the canonical transcript
```{r fig.height=5, fig.width=11}
#try compare with ref
can_exons <- my_id_gtf %>% filter(transcript_id == "CLN3_24_438aa_5UTR_136_3UTR_49") %>% filter(type == "exon")


trans_info <- trans_merged_filtered %>% filter(orf_id == "CLN3_235_181aa")
  trans_id <- trans_info$new_id
  major_gtf <- my_id_gtf_filtered %>% filter(transcript_id %in% trans_id)
  major_cds <- my_id_cds_filtered %>% filter(transcript_id %in% trans_id)

major_diffs <- to_diff(
  exons = major_gtf,
  ref_exons = can_exons,
  group_var = "transcript_id"
)


major_gtf %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = transcript_id,
  
)) +
  geom_range(
    # fill = "white",
    height = 0.25
  ) +
  geom_range(
    data = major_cds,
    aes()
  ) +
  geom_intron(
    data = to_intron(major_gtf, "transcript_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + 
  theme_bw()+ 
  xlab("Coordinates on CHR16") + 
  ylab(NULL)+
  geom_range(
    data = major_diffs,
    aes(fill = diff_type),
    alpha = 0.2,
    linetype = 2
  ) + theme(axis.title = element_text(face = "bold"), 
            legend.position = "bottom") + 
  ggtitle('Comparison of transcripts with the same ORF as the major disease transcript') + 
    scale_fill_discrete(
      name = paste0("Region in canonical transcript:"), 
      labels = c(
        paste0("In canonical transcript"), 
        paste0("Not in canonical transcript")
      )) +
    scale_colour_discrete(
      name = paste0("Region in canonical transcript:"), 
      labels = c(
        paste0("In canonical transcript"), 
        paste0("Not in canonical transcript")
      )
    ) 

  


```

Trnascript usage across all tissues and tissue-specificity

```{r fig.height=32.5, fig.width=20}
# Generate expression heat map
exp_heat <- ggplot(tissue_usage, aes(tissue, reorder(new_id, occurrence))) +
  geom_tile(aes(fill = avg_prop), color = NA) +
  scale_fill_viridis(labels = scales::label_percent(accuracy = 0.1)) +
  facet_grid(. ~ organ, scales = "free", space = "free") +
  theme_bw() +
  xlab("Tissues") +
  theme(
    panel.border = element_blank(),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 11),
    axis.text.y = element_text(size = 11),
    strip.text.x = element_text(margin = margin(0.5, 0, 0.5, 0, "cm"), angle = 90),
    strip.text = element_text(face = "bold", size = rel(1.2)),
    axis.title.x = element_text(face = "bold", size = 12)
  ) +
  guides(fill = guide_legend(title = "Usage"))



my_id_gtf_filtered$ts <- tsinfo$tissue_specific[match(my_id_gtf_filtered$transcript_id, tsinfo$new_id)]
my_id_cds_filtered$ts <- tsinfo$tissue_specific[match(my_id_cds_filtered$transcript_id, tsinfo$new_id)]

# Generate structure plots
structure_filtered_ts <- my_id_gtf_filtered %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = reorder(transcript_id, occurrence)
)) +
  geom_range(height = 0.25) +
  geom_range(data = my_id_cds_filtered, aes(fill = ts)) +
  geom_intron(
    data = to_intron(my_id_gtf_filtered, "transcript_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500
  ) +
  theme_bw() +
  xlab("Coordinates on CHR16") +
  ylab(NULL) +
  guides(fill = guide_legend(ncol = 1)) +
  scale_fill_manual("Tissue-specific", values = c("TRUE" = "Orange", "FALSE" = "Black")) +
  theme(axis.title.x = element_text(face = "bold"))

# Generate structure plots for top transcripts
structure_filtered_ts_top <- my_id_gtf_top15 %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = reorder(transcript_id, occurrence)
)) +
  geom_range(height = 0.25) +
  geom_range(data = my_id_cds_top15, aes(fill = ts)) +
  geom_intron(
    data = to_intron(my_id_gtf_top15, "transcript_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500
  ) +
  theme_bw() +
  xlab("Coordinates on CHR16") +
  ylab(NULL) +
  guides(fill = guide_legend(ncol = 1)) +
  scale_fill_manual("Tissue-specific", values = c("TRUE" = "Orange", "FALSE" = "Black")) +
  theme(axis.title.x = element_text(face = "bold"))


plot1 <- cowplot::plot_grid(
  exp_heat + 
    theme(axis.title.y = element_blank(), 
          legend.position = "left"),
  structure_filtered_ts + 
    theme(axis.text.y = element_blank(), 
          axis.ticks.y = element_blank(), 
          axis.title.y = element_blank()),
  nrow = 1, 
  labels = "AUTO", 
  rel_widths = c(1.8, 2), 
  align = "h", 
  axis = "bt"
)



#Show only top 15 transcripts 



result_top <- tissue_usage %>% filter(new_id %in% my_id_cds_top15$transcript_id)
exp_heat_top <- ggplot(result_top, aes(tissue, reorder(new_id, occurrence))) +
  geom_tile(aes(fill = avg_prop), color = NA) +
  scale_fill_viridis(begin = 0, end = 1, labels = scales::label_percent(accuracy = 0.1)) +
  facet_grid(. ~ organ, scales = "free", space = "free") +
  theme_bw() +
  xlab("Tissues") +
  theme(
    panel.border = element_blank(),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 11),
    axis.text.y = element_text(size = 11),
    strip.text.x = element_text(margin = margin(0.5, 0, 0.5, 0, "cm"), angle = 90),
    strip.text = element_text(face = "bold", size = rel(1.2)),
    axis.title.x = element_text(face = "bold", size = 12)
  ) +
  guides(fill = guide_legend(title = "Usage"))


#plot structure for top transcripts

my_id_gtf_top15$ts <- tsinfo$tissue_specific[match(my_id_gtf_top15$transcript_id, tsinfo$new_id)]
my_id_cds_top15$ts <- tsinfo$tissue_specific[match(my_id_cds_top15$transcript_id, tsinfo$new_id)]

structure_filtered_ts_top <- my_id_gtf_top15 %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = reorder(transcript_id, occurrence),
  
)) +
  geom_range(
    # fill = "white",
    height = 0.25
  ) +
  geom_range(
    data = my_id_cds_top15,
    aes(fill = ts)
  ) +
  geom_intron(
    data = to_intron(my_id_gtf_top15, "transcript_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + theme_bw()+ xlab("Coordinates on CHR16") + ylab(NULL) + 
  guides(fill = guide_legend(ncol = 1))+ 
  scale_fill_manual("Tissue-specific",values = c("TRUE" = "Orange", "FALSE" = "Black")) + 
  theme(axis.title.x = element_text(face = "bold"))


plot2 <- cowplot::plot_grid(
  exp_heat_top + 
    theme(axis.title.y = element_blank(), 
          legend.position = "left"),
  structure_filtered_ts_top + 
    theme(axis.text.y = element_blank(), 
          axis.ticks.y = element_blank(), 
          axis.title.y = element_blank()),
  nrow = 1, 
  labels = "AUTO", 
  rel_widths = c(1.8, 2), 
  align = "h", 
  axis = "bt"
)


plot1
plot2
```



Visualise alternative UTRs
```{r fig.height=6.3, fig.width=15}

plot_utr("CLN3_235_181aa", 3)

info_181 <- utr_ts_info
```


```{r fig.height=8.5, fig.width=15}
plot_utr("CLN3_207_223aa", 5)

info_223 <- utr_ts_info
```


```{r fig.height=10, fig.width=15}
plot_utr("CLN3_24_438aa", 5)

info_438 <- utr_ts_info
```

```{r fig.height=5.5, fig.width=15}
plot_utr("CLN3_24_438aa", 3)

info_438 <- utr_ts_info
```


Compare ORFs with novel seqs with canonical ORF
```{r fig.height=4, fig.width=15}
#CLN3_80_363
#CLN3_81_363
#one with extra exon, the other one has alternative start codon in coding exon 3 of canonical transcript

trans_info <- info %>% filter(orf_id %in% c("CLN3_4_489aa", "CLN3_125_316aa", "CLN3_101_338aa", "CLN3_80_363aa", "CLN3_235_181aa", "CLN3_115_328aa"))
  trans_id <- trans_info$new_id
  
  orfcds <- my_id_cds %>% filter(transcript_id %in% trans_id)

  orfcds$orf_id <- info$orf_id[match(orfcds$transcript_id, info$new_id)]
  
  orfcds <- orfcds %>% dplyr::select(-c(transcript_id, occurrence, transcript_novelty, transcript_status)) %>% unique()

  

 can_exons <- my_id_cds %>% filter(transcript_id == "CLN3_24_438aa_5UTR_136_3UTR_49") %>% filter(type == "CDS")



orf_diffs <- to_diff(
  exons = orfcds,
  ref_exons = can_exons,
  group_var = "orf_id"
)

orf_dif <- orfcds %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = orf_id,
  
)) +
  geom_range(
    # fill = "white",
    height = 0.35
  ) +
  geom_intron(
    data = to_intron(orfcds, "orf_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + theme_bw()+ xlab("Coordinates on CHR16") + ylab(NULL)+
  geom_range(
    data = orf_diffs,
    aes(fill = diff_type),
    alpha = 0.2,
    linetype = 2
    
  ) + theme(axis.title = element_text(face = "bold"), 
            legend.position = "bottom", axis.text.y = element_text(size = 15)
            ) + 
      scale_fill_discrete(
      name = paste0("Region in canonical ORF:"), 
      labels = c(
        paste0("In canonical ORF"), 
        paste0("Not in canonical ORF")
      )) +
    scale_colour_discrete(
      name = paste0("Region in canonical ORF:"), 
      labels = c(
        paste0("In canonical ORF"), 
        paste0("Not in canonical ORF")
      )
    ) 


orf_dif
```


Show 438aa ORF, combine with ORF differences
```{r fig.height=5, fig.width=14}
trans_info <- info %>% filter(orf_id %in% c("CLN3_24_438aa"))
  trans_id <- trans_info$new_id
  #orfgtf <- my_id_gtf_filtered %>% filter(transcript_id %in% trans_id)
  orfcds <- my_id_cds %>% filter(transcript_id %in% trans_id)

  orfcds$orf_id <- info$orf_id[match(orfcds$transcript_id, info$new_id)]
  
  orfcds <- orfcds %>% dplyr::select(-c(transcript_id, occurrence, transcript_novelty, transcript_status)) %>% unique()
  
  can_orf <- orfcds %>% ggplot(aes(
  xstart = start,
  xend = end,
  y = orf_id,
  fill = orf_id
)) +
  geom_range(
    # fill = "white",
    height = 0.30
  ) +
  geom_intron(
    data = to_intron(orfcds, "orf_id"),
    aes(strand = strand),
    arrow.min.intron.length = 500,
  ) + 
    theme_bw()+ 
    xlab("Coordinates on CHR16") + 
    ylab(NULL) + 
    scale_fill_manual(values = color_used)
    
  
  
  #can_orf
  
  cowplot::plot_grid(can_orf + theme(legend.position = "none", 
                                     axis.text.x = element_blank(),
                                     axis.title.x = element_blank(),
                                     axis.text.y = element_text(size = 14)), orf_dif, rows = 2, rel_heights = c(1.5,6), align = "v")
```


Visualise alignment for peptides and corresponding proteins
```{r fig.height=6, fig.width=20}
library(ggmsa)

ggmsa(msa = "C:/Users/Haoyu/Desktop/MSsearch/489_pep_align.fasta", seq_name = T, char_width = 0.7) + theme(axis.text = element_text(size = 14))
ggmsa(msa = "C:/Users/Haoyu/Desktop/MSsearch/major_pep_align.fasta", seq_name = T, char_width = 0.7) + theme(axis.text = element_text(size = 28))
```
